### PR TITLE
Restructuring ModelVersion and LocalModel 

### PR DESCRIPTION
**Pull request by tidylobster**
* Introduced breaking changes to the API: class names, method names, method parameters, etc.
* Introduced unified request error handling for checking/raising client-side/server-side errors.
* Removed all code for parsing yaml definitions. It only makes sense to use it within CLI, so it will reside there.
* Removed UploadResponse object. Now LocalModel.upload returns ModelVersion instance directly.
* Removed Metricable class as it's no longer makes sense to carry arround a whole class for just two methods for a signle usage. 
* Removed Cluster.connect method in favor of pure class constructor. 
* Took out uploading metrics from LocalModel and transferred functionality to ModelVersion. Now there are two new methods on a ModelVersion instance: `.as_metric(threshold=10, comparator: ThresholdCmpOp.LESS_EQ)` and `.assign_metrics([metric])`. Instead of manually handling MetricSpecConfig, MetricSpec user can work with this:

    ```python
    mv1 = ModelVersion.find_by_id(cluster, 1)
    mv2 = ModelVersion.find_by_id(cluster, 2)
    mv2.assign_metrics([mv1.as_metric(10, ThresholdCmpOp.LESS_EQ)])
    ```
* Added functionality for uploading training data. Current implementation of `wait` is a subject of change in the future releases.

    ```python
    modelversion = ModelVersion.find(cluster, "my-model", 1)
    modelversion.training_data = "s3://bucket/path/to/training-data.csv"
    data_upload_response = modelversion.upload_training_data()
    data_upload_response.wait(retry=3, sleep=30)
    ```
* Improved/refactored tests regarding cluster entities (Application, ModelVersion, etc). 
* Added/updated class-level documentation with code-snippets for important Hydrosphere entities. 
* Fixed various misspellings and minor bugs.

### CODE REVIEW

File: hydrosdk/modelversion.py [hunk 1]
```diff
@@ -535,11 +593,80 @@ class NotFound(Exception):
...
+
+    class DataProcessingNotFinished(Exception):
+        pass
+
+    class DataProcessingFailed(Exception):
         pass
 
 
-class UploadResponse:
+class ModelUploadResponse:
```
[the remark this answers is not in the document]
Comment (Valenzione) [reply hydrosdk/modelversion.py:669]: Personally, I am in favor of this renaming. Despite extending a name by a few characters it prevents any possible misunderstanding.

[the remark this answers is not in the document]
Comment (trthhrtz) [reply hydrosdk/modelversion.py:669]: @Valenzione it creates ambiguity. At the very least there should be a hint in the docstring. Otherwise imho the name should be associated either with localModel or modelVersion as we don't have a model entity

File: hydrosdk/modelversion.py [hunk 2]
```diff
@@ -535,11 +593,80 @@ class NotFound(Exception):
     class BadRequest(Exception):
         pass
 
-    class BadResponse(Exception):
```
[the remark this answers is not in the document]
Comment (tidylobster) [reply hydrosdk/modelversion.py:538]: BadResponse can appear not only for LocalModel/ModelVersion but for DataUpload as well. In my mind those are 5xx requests

File: hydrosdk/modelversion.py [hunk 3]
```diff
@@ -126,10 +176,10 @@ class LocalModel(Metricable):
     """
 
     @staticmethod
-    def create(path, name, runtime, contract=None, install_command=None):
+    def create(name, runtime, payload, contract=None, install_command=None):
```
[the remark this answers is not in the document]
Comment (tidylobster) [reply hydrosdk/modelversion.py:179]: Yes, I've discussed this with @KineticCookie, this method should only be used for internal usage. E.g. when we just want to create an instance of LocalModel without reading serving.yaml, but specifying all fields ourselves. Need to create an issue for this one.

File: hydrosdk/modelversion.py [hunk 4]
```diff
@@ -489,6 +540,11 @@ def list_modelversions_by_model_name(cluster: Cluster, model_name: str) -> list:
         sorted_by_version = sorted(modelversions_by_name, key=lambda model: model.version)
 
         return sorted_by_version
+    
+    def upload_training_data(self) -> 'DataUploadResponse':
```
[the remark this answers is not in the document]
Comment (tidylobster) [reply hydrosdk/modelversion.py:544]: Data upload operates only on the modelversion, it doesn't make sense to utilize in LocalModel class

File: hydrosdk/modelversion.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Valenzione) [reply hydrosdk/modelversion.py:669]: Ah, yeah, I missed that point, I got focused on `Why renaming?`. I agree with the ambiguity mentioned.

[the remark this answers is not in the document]
Comment (tidylobster) [reply hydrosdk/modelversion.py:669]: I've introduced a new DataUploadResponse class, which means having UploadResponse might be confusing. ModelUploadResponse makes it more verbose

File: hydrosdk/modelversion.py [hunk 5]
```diff
@@ -20,6 +21,51 @@
...
+def _upload_local_file(cluster: Cluster, model_version_id: int, path: str, chunk_size=1024):
+    """
+    Internal method for uploading local training data to Hydrosphere.
+
+    :param cluster: Cluster instance
+    :param url: url to which upload the file
+    :param path: path to a local file
+    :param chunk_size: chunk size to use for streaming
+    """
+    def read_in_chunks(filename, chunk_size):
```
[the remark this answers is not in the document]
Comment (tidylobster) [reply hydrosdk/modelversion.py:43]: Well, I thought, since we are utilizing this function only within this method, it would be okay to have it as a subfunction

File: hydrosdk/modelversion.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tidylobster) [reply hydrosdk/modelversion.py:538]: BadRequest (in the ModelVersion) on the other hand might signal that there's something wrong with the model definition, so server will return 4xx request.

File: hydrosdk/modelversion.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (trthhrtz) [reply hydrosdk/modelversion.py:547]: Okay, but why you did not do this:

```
        if self.training_data is not None:
            return upload_training_data(self.cluster, self.id, self.training_data)
        raise ValueError('Training data is not specified')
```
:)

[the remark this answers is not in the document]
Comment (tidylobster) [reply hydrosdk/modelversion.py:547]: Do this where?

[the remark this answers is not in the document]
Comment (trthhrtz) [reply hydrosdk/modelversion.py:547]: Ah, ok. I see your point. 

Moving forward, - 

1) I thought you said that both modelversion and localmodel will have these methods?

2) As a developer for me it's hard to understand the difference between 2 upload_training_data methods. When which should be used and which of them is internal?

File: hydrosdk/modelversion.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (trthhrtz) [reply hydrosdk/modelversion.py:672]: @tidylobster it's just that we don't have a Model entity and using it anywhere makes an ambiguity...

File: hydrosdk/exceptions.py [hunk 6]
```diff
@@ -1,3 +1,21 @@
+class BadRequest(BaseException):
+    """
+    Exception for client-side errors.
+    """
+    pass
+
+
+class BadResponse(BaseException):
```
Comment (Valenzione) [inline hydrosdk/exceptions.py:8]: Python [documentation](https://docs.python.org/3/library/exceptions.html) states that 
>The built-in exception classes can be subclassed to define new exceptions; programmers are encouraged to derive new exceptions from the `Exception` class or one of its subclasses, and **not from** `BaseException`. More information on defining exceptions is available in the Python Tutorial under User-defined Exceptions.

I suggest following this guideline

**Comment by tidylobster** [comment]
What has been done: 
* Introduced breaking changes to the API, hense class names, method names, method parameters, etc.
* Introduced unified request error handling for checking/raising client-side/server-side errors.
* Removed all code for parsing yaml definitions. It only makes sense to use it within CLI, so it will reside there.
* Removed UploadResponse object. Now LocalModel.upload returns ModelVersion instance directly.
* Removed Metricable class as it's no longer makes sense to carry arround a whole class for just two methods for a signle usage. 
* Removed Cluster.connect method in favor of pure class constructor. 
* Took out uploading metrics from LocalModel and transferred functionality to ModelVersion. Now there are two new methods on a ModelVersion instance: `.as_metric(threshold=10, comparator: ThresholdCmpOp.LESS_EQ)` and `.assign_metrics([metric])`. Instead of manually handling MetricSpecConfig, MetricSpec user can work with this:

    ```python
    mv1 = ModelVersion.find_by_id(cluster, 1)
    mv2 = ModelVersion.find_by_id(cluster, 2)
    mv2.assign_metrics([mv1.as_metric(10, ThresholdCmpOp.LESS_EQ)])
    ```
* Added functionality for uploading training data. Current implementation of `wait` is a subject of change in the future releases.

    ```python
    modelversion = ModelVersion.find(cluster, "my-model", 1)
    modelversion.training_data = "s3://bucket/path/to/training-data.csv"
    data_upload_response = modelversion.upload_training_data()
    data_upload_response.wait(retry=3, sleep=30)
    ```
* Improved/refactored tests regarding cluster entities (Application, ModelVersion, etc). 
* Added/updated class-level documentation with code-snippets for important Hydrosphere entities. 
* Fixed various misspellings and minor bugs.

**Comment by tidylobster** [comment]
What has been done: 
* Introduced breaking changes to the API, hense class names, method names, method parameters, etc.
* Introduced unified request error handling for checking/raising client-side/server-side errors.
* Removed all code for parsing yaml definitions. It only makes sense to use it within CLI, so it will reside there.
* Removed UploadResponse object. Now LocalModel.upload returns ModelVersion instance directly.
* Removed Metricable class as it's no longer makes sense to carry arround a whole class for just two methods for a signle usage. 
* Removed Cluster.connect method in favor of pure class constructor. 
* Took out uploading metrics from LocalModel and transferred functionality to ModelVersion. Now there are two new methods on a ModelVersion instance: `.as_metric(threshold=10, comparator: ThresholdCmpOp.LESS_EQ)` and `.assign_metrics([metric])`. Instead of manually handling MetricSpecConfig, MetricSpec user can work with this:

    ```python
    mv1 = ModelVersion.find_by_id(cluster, 1)
    mv2 = ModelVersion.find_by_id(cluster, 2)
    mv2.assign_metrics([mv1.as_metric(10, ThresholdCmpOp.LESS_EQ)])
    ```
* Added functionality for uploading training data. Current implementation of `wait` is a subject of change in the future releases.

    ```python
    modelversion = ModelVersion.find(cluster, "my-model", 1)
    modelversion.training_data = "s3://bucket/path/to/training-data.csv"
    data_upload_response = modelversion.upload_training_data()
    data_upload_response.wait(retry=3, sleep=30)
    ```
* Improved/refactored tests regarding cluster entities (Application, ModelVersion, etc). 
* Added/updated class-level documentation with code-snippets for important Hydrosphere entities. 
* Fixed various misspellings and minor bugs.

File: tests/common_fixtures.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+
+@pytest.fixture(scope="session")
+def payload():
+    return ['./src/func_main.py']
+
+
+@pytest.fixture(scope="session")
+def runtime():
+    return DockerImage("hydrosphere/serving-runtime-python-3.6", "2.1.0", None)
```
Comment (Valenzione) [inline tests/common_fixtures.py:38]: Why `2.1.0` and not `2.3.2`?

File: hydrosdk/utils.py [hunk 8]
```diff
@@ -1,18 +1,56 @@
-import grpc
+from typing import Generator
 
-ALLOWED_TIMEOUT_gRPC_SECONDS = 5
```
Comment (trthhrtz) [inline hydrosdk/utils.py:3]: why deleting default timeout?

File: tests/common_fixtures.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
+import os
+
+import pytest
+import grpc
+
+from hydrosdk.cluster import Cluster
+from hydrosdk.image import DockerImage
+from hydrosdk.contract import ModelContract, SignatureBuilder
+from hydrosdk.modelversion import LocalModel
+from tests.config import *
```
Comment (trthhrtz) [inline tests/common_fixtures.py:10]: > Wildcard imports (from <module> import *) should be avoided, as they make it unclear which names are present in the namespace, confusing both readers and many automated tools. 
https://www.python.org/dev/peps/pep-0008/#imports

File: tests/test_servable.py [hunk 10]
```diff
@@ -1,89 +1,81 @@
...
+    mv.lock_till_released()
+    sv: Servable = Servable.create(cluster, mv.name, mv.version)
+    assert Servable.find_by_name(cluster, sv.name)
+
+
+def test_servable_list_all(cluster: Cluster, local_model: LocalModel):
+    mv: ModelVersion = local_model.upload(cluster)
+    mv.lock_till_released()
+    sv: Servable = Servable.create(cluster, mv.name, mv.version)
+    assert sv.name in [servable.name for servable in Servable.list_all(cluster)]
```
Comment (trthhrtz) [inline tests/test_servable.py:25]: error messages in case of assert is wrong would be helpful

**Comment by trthhrtz** [comment]
Great work 👍 . 


p.s. A bit difficult to code-review this large PR's 😄 

@KineticCookie  what do you think about
> Removed all code for parsing yaml definitions. It only makes sense to use it within CLI, so it will reside there.
?

File: tests/test_servable.py (hunk 10 above)
Comment (Valenzione) [reply tests/test_servable.py:25]: @trthhrtz Would you mind suggesting changes while @tidylobster is on vacation?

File: hydrosdk/application.py [hunk 11]
```diff
@@ -22,135 +22,177 @@ class ApplicationStatus(Enum):
...
+    def _from_json(cluster: Cluster, application_json: dict) -> 'Application':
+        """
+        Deserialize json into application object. 
+
+        :param cluster: active cluster
+        :param application_json: input json with application object fields
+        :return: application object
+        """
+        app_id = application_json.get("id")
+        app_name = application_json.get("name")
```
[the remark this answers is not in the document]
Comment (Valenzione) [reply hydrosdk/application.py:103]: I think there are no such cases, wdyt @KineticCookie ?

File: hydrosdk/cluster.py [hunk 12]
```diff
@@ -68,22 +59,19 @@ def __init__(self, http_address: str,
...
                     raise ValueError("Missing grpc credentials")
-
                 self.channel = grpc.secure_channel(target=self.grpc_address, credentials=grpc_credentials,
                                                    options=grpc_options,
                                                    compression=grpc_compression)
             else:
                 self.channel = grpc.insecure_channel(target=self.grpc_address, options=grpc_options,
                                                      compression=grpc_compression)
 
-            # with grpc.secure_channel it takes a lot of time to check the grpc-connection
```
[the remark this answers is not in the document]
Comment (Valenzione) [reply hydrosdk/cluster.py:79]: Is this error a reproducible one to leave it here?

File: hydrosdk/utils.py (hunk 8 above)
Comment (Valenzione) [reply hydrosdk/utils.py:3]: Because @tidylobster have moved timeout to the `Cluster` constructor

**Comment by Valenzione** [comment]
> Great work 👍 .


@trthhrtz Would you mind changing your review status from "Request Changes" to "Approved" then?) 🙂

File: hydrosdk/application.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (trthhrtz) [reply hydrosdk/application.py:103]: @Valenzione If there are none I see it to be ambiguous sometimes to use `.get()` and sometimes to access dict's elements directly

**Comment by trthhrtz** [comment]
@Valenzione yeah, I just need to get replies from @KineticCookie and ready to approve

**Comment by trthhrtz** [comment]
Did minor refactoring, resolved conflicts @Valenzione

**Comment by trthhrtz** [comment]
👍

File: hydrosdk/application.py [hunk 13]
```diff
@@ -22,135 +22,177 @@ class ApplicationStatus(Enum):
...
+    :Example:
+    
+    Create an application from existing modelversions.
+    >>> from hydrosdk.cluster import Cluster
+    >>> from hydrosdk.modelversion import ModelVersion
+    >>> cluster = Cluster('http-cluster-endpoint')
+    >>> mv1 = ModelVersion.find(cluster, "my-model", 1)
+    >>> mv2 = ModelVersion.find(cluster, "my-model", 2)
+    >>> stage = ExecutionStageBuilder() \
+                    .with_model_variant(mv1, 50) \
```
[the remark this answers is not in the document]
Comment (tidylobster) [reply hydrosdk/application.py:174]: Every stage can be comprized of multiple model versions. We call them model variants only within stages.

File: hydrosdk/contract.py [hunk 14]
```diff
@@ -449,10 +386,20 @@ def build(self) -> ModelSignature:
             inputs=self.inputs,
             outputs=self.outputs
         )
-
-    def __with_field(self, collection, name, dtype, shape, profile=ProfilingType.NONE):
+                    
+    def __with_field(self, collection: List[ModelField], name: str,
```
[the remark this answers is not in the document]
Comment (tidylobster) [reply hydrosdk/contract.py:390]: Left this as it was before. We have to make a convention regarding single/double underscore notations.
